### ISSUE TITLE
Negative indexes with end of buffer semantics

### ISSUE THREAD

**Issue by lizmat** [MoarVM/MoarVM#1024]
Sometime in the past two days, the read-int/write-int functions suddenly allowed negative indices, and gave them end-of-buffer semantics: -1 = last byte of buffer, -2 the one before that, etc.

    my $b = buf8.new(1..2); $b.write-int8(-1,42); say $b   # Buf[uint8]:0x<01 2A>

Is this intentional?  If so, then we'll have to burden the setting with a HLL check for negative values for indices.  If not, then this is a bug that needs fixing :-)

**Comment by niner** [comment]
Yes, it was intentional to match all the other VMArray operations. I guess since the situation is the same with plain Buf indexing, we have a somewhat performant solution?

**Comment by lizmat** [comment]
I guess if we do things often enough, the JIT will DTRT.  But still, this feels like cargo culting a bad decision from the past into newly created code.  If I look at how little code is needed at the VM level to handle this situation and compare it to the amount of code and bloat that is needed at the HLL level, I wonder why we continue to do it like this.  I don't think there's any code in the VM itself that uses these ops, let alone using negative indices on them.

**Comment by jnthn** [comment]
Given MoarVM exists to be a VM for Perl 6, I don't really feel we should go on providing the negative wrap-around semantics in the VM. I'd be happy with just treating `idx < 0` as an error everywhere, and making anyone who wants the wrap-around semantics building on MoarVM pay for the code in HLL space to do it.

So far as I remember, the `VMArray` semantics came from alignment with the Parrot `ResizablePMCArray` ones, to ease the process of porting NQP and Rakudo to MoarVM. And of course, Parrot was meant to support many languages, so it was quite defensible to have those semantics since some target languages would want them. With MoarVM that's pretty much been an ungoal, so we can just do what makes things easiest for Perl 6 compiler implementation.

**Comment by lizmat** [comment]
FWIW, I've added the checks in the setting: https://github.com/rakudo/rakudo/commit/5f3a955aed , hopefully to be reverted soon

**Comment by niner** [comment]
Glad to hear that :)
I do wonder however, that if negative values aren't allowed, why is $offset an int and not a uint?

When using these ops, what I'd have wished for was not to easily write to the last element of the buf, but to be able to append without having to keep track of the current position. An offset of -1 could be used for that. Or, since we have a $flags argument already, we could just add a flag for how the offset should be interpreted.
